Samples/Widgets/SingleWidget/cpp-win32-packaged/SampleWidgetProviderApp/main.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "WidgetProvider.h"
+#include <mutex>
+
+// This GUID is the same GUID thas was provided in the 
+// registration of the COM Server and Class Id in the .appxmanifest.
+static constexpr GUID widget_provider_clsid
+{ /* 8AFD9DB6-84A7-4425-BD7E-761D9C063B40 */
+    0x8afd9db6, 0x84a7, 0x4425, { 0xbd, 0x7e, 0x76, 0x1d, 0x9c, 0x6, 0x3b, 0x40 }
+};
+
+// This is implementation of a ClassFactory that will instantiate WidgetProvider
+// that will interact with the Widget Service.
+template <typename T>
+struct SingletonClassFactory : winrt::implements<SingletonClassFactory<T>, IClassFactory>
+{
+    STDMETHODIMP CreateInstance(
+        ::IUnknown* outer,
+        GUID const& iid,
+        void** result) noexcept final
+    {
+        *result = nullptr;
+
+        std::unique_lock lock(mutex);
+
+        if (outer)
+        {
+            return CLASS_E_NOAGGREGATION;
+        }
+
+        if (!instance)
+        {
+            instance = winrt::make_self<T>();
+        }
+
+        return instance.as(iid, result);
+    }
+
+    STDMETHODIMP LockServer(BOOL) noexcept final
+    {
+        return S_OK;
+    }
+
+private:
+    winrt::com_ptr<T> instance{ nullptr };
+    std::mutex mutex;
+};
+
+int main()
```

Comment:
When will this process be shut down? Does it have a termination heuristic?  See also https://learn.microsoft.com/windows/win32/com/out-of-process-server-implementation-helpers and linked items.

Reply:
it doesn't have termination heuristic. It's a good idea to add it. Once no outstanding references are alive - we can terminate.

Reply:
The widget provider will need some sort of new API for this, right? There's no way I'm aware of for a provider to know it's safe to shutdown.

Reply:
You might be right - I'll need to read up on the inners of out of proc COM and whether provider can see if any of cocreated objects are still alive. Otherwise, without new API there's no way:)

Reply:
@jonwis @riverar - although there's no API available to signal the provider that it can shut down - we can implement general shutdown on idle. It will look like this: (I'll add to the sample)
```
wil::unique_event g_shudownEvent(wil::EventOptions::None);

void SignalServerShutdown()
{
    g_shudownEvent.SetEvent();
}

namespace winrt
{
    inline auto get_module_lock() noexcept
    {
        struct service_lock
        {
            uint32_t operator++() noexcept
            {
                auto ref = ::CoAddRefServerProcess();
                std::cout << __FUNCTION__ " ServerLockCount:" << ref << std::endl;
                return ref;
            }

            uint32_t operator--() noexcept
            {
                auto ref = ::CoReleaseServerProcess();
                std::cout << __FUNCTION__ " ServerLockCount:" << ref << std::endl;
                if (ref == 0)
                {
                    std::cout << "SignalServerShutdown()" << std::endl;
                    SignalServerShutdown();
                }
                return ref;
            }
        return service_lock{};
    }
}

int main()
{
    DWORD index{};
    HANDLE events[] = {g_shudownEvent.get()};
    winrt::check_hresult(CoWaitForMultipleObjects(CWMO_DISPATCH_CALLS | CWMO_DISPATCH_WINDOW_MESSAGES, INFINITE, static_cast<ULONG>(std::size(events)), events, &index));
return 0;
```

Reply:
the logic for shutdown has been added.